### PR TITLE
 Remove dependencies on curses/terminfo/termcap C library

### PR DESCRIPTION
**Pull request by xavierleroy**
The toplevel REPL uses terminal escape sequences to display error messages nicely, underlining the location of the error when possible.  Currently, this hack with the terminal goes through the `tgetent`, `tgetstr`, `tgetnum` and `tputs` functions provided by one of the curses, terminfo or termcap C libraries.  The standard bytecode runtime system provides OCaml bindings to these functions, and therefore depends on the curses/terminfo/termcap C library used.

This pull request breaks this dependency by
* Using ANSI escape sequences instead of querying a termcap/terminfo database.  The days of pre-ANSI terminals is gone; now all consoles of interest implement ANSI escape sequences.  Besides, for colored error messages, the OCaml compilers already use ANSI escape sequences.
* Using an `ioctl` to query the number of lines on the screen.  No extra C library is needed for this.  A reasonable default of 24 is used if the  `ioctl` fails or is not available (Win32).

This PR requires a bootstrap (because some primitives are removed).  Actually two bootstraps are needed because the boostrapping procedure is somewhat broken currently.  The first commit in this PR adds the new primitive but keeps the old ones.  The second commit is the result of the double bootstrap with removal of the old primitives in between.  I am willing to do the merge and redo the bootstrap "live" when this PR is accepted.

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -152,7 +151,7 @@ while : ; do
     -lib*)
         cclibs="$2 $cclibs"; shift;;
     -no-curses|--no-curses)
-        with_curses=no;;
+        ;;  # Ignored for backward compatibility
```
Comment (dbuenzli) [inline configure:154]: A warning that this has no effet could be `echo`ed.

File: utils/terminfo.ml [hunk 2]
```diff
@@ -13,14 +13,35 @@
...
-external backup : int -> unit = "caml_terminfo_backup";;
-external standout : bool -> unit = "caml_terminfo_standout";;
-external resume : int -> unit = "caml_terminfo_resume";;
+  | Good_term of int  (* number of lines of the terminal *)
+
+let setup () = 
+  let term = try Sys.getenv "TERM" with Not_found -> "" in
+  (* Same heuristics as in Misc.Color.should_enable_color *)
+  if term <> "" && term <> "dumb" && isatty stderr then begin
+    let rows = terminfo_rows stderr in
```
Comment (dbuenzli) [inline utils/terminfo.ml]: Previously the `setup` analysis was done on `stdout` as per the given argument. You now do this on `stderr` is that change intended  ?

**Comment by shindere** [comment]
What a nice idea! Thanks!

What exactly is broken in the boostrap procedure?

Regarding the default values when the ioctl fails, how about also taking
into account environment variables such as LINES and COLUMNS before
falling back to hard-coded defaults?

**Comment by nojb** [comment]
Nice! IIUC, this does not account for window resizing, but one could use the `SIGWINCH` signal for that.

**Comment by avsm** [comment]
Just a thank you note for this PR, since it removes yet more C stubs from the embedded unikernel builds of MirageOS :-)

File: utils/terminfo.ml (hunk 2 above)
Comment (xavierleroy) [reply utils/terminfo.ml]: The change was for consistency with `Misc.Color.should_enable_color`, which uses `stderr`.

However, a point could be made that the latter is oriented towards compiler error messages (on stderr) while the present code is targeted at the toplevel REPL error messages (on stdout, if I'm not mistaken).

I don't have a strong feeling either way.

File: configure (hunk 1 above)
Comment (xavierleroy) [reply configure:154]: I just followed the pattern of the `-with-pthread*` ignored option.

**Comment by xavierleroy** [comment]
> this does not account for window resizing

I'm pretty sure the `ioctl` does take window resizing into account.  So, it would be enough to stop cacheing the result of `Terminfo.setup` in `Location.highlight_locations`.  Then the `ioctl` would run every time an error message needs highlighting.

Added 2017-10-17: commit 88c2987 does just what I outlined above.

**Comment by xavierleroy** [comment]
> Regarding the default values when the ioctl fails, how about also taking into account environment variables such as LINES and COLUMNS before falling back to hard-coded defaults?

These variables are defined by shells but not exported to processes.
```
$ echo $LINES
24
$ echo 'Sys.getenv "LINES";;' | ocaml
        OCaml version 4.06.0+beta1

# Exception: Not_found.
```

**Comment by xavierleroy** [comment]
> What exactly is broken in the boostrap procedure?

I'll submit a Mantis bug report, eventually.  In a nutshell: executables generated by `./ocamlc` or by `boot/ocamlc -use-prims byterun/primitives` (as opposed to plain `boot/ocamlc`) should be executed by `byterun/ocamlrun` and not by `boot/ocamlrun`, because they may use primitives that are provided by the former but not by the latter.

**Comment by shindere** [comment]
Xavier Leroy (2017/10/16 10:26 -0700):
> The change was for consistency with `Misc.Color.should_enable_color`, which uses `stderr`.
> 
> However, a point could be made that the latter is oriented towards compiler error messages (on stderr) while the present code is targeted at the toplevel REPL error messages (on stdout, if I'm not mistaken).
> 
> I don't have a strong feeling either way.

Would it make sense to make the output configurable somehow?

Then, no matter the default, it would not hurt at all.

**Comment by shindere** [comment]
Xavier Leroy (2017/10/16 17:32 +0000):
> > Regarding the default values when the ioctl fails, how about also taking into account environment variables such as LINES and COLUMNS before falling back to hard-coded defaults?
> 
> These variables are defined by shells but not exported to processes.

OK. I guess my concern is that I'd like to have a way to be able to use
my good old 80x25 terminal.

File: utils/terminfo.ml (hunk 2 above)
Comment (dbuenzli) [reply utils/terminfo.ml]: I'd say it's better to invoke `isatty` and `terminfo_rows` on the fd that you are actually going to use for output and in this case it seems to be stdout:

```
> ocaml -noinit 2> /dev/null
        OCaml version 4.03.0

# let f x = let;;
Error: Syntax error
```

Comment (xavierleroy) [reply utils/terminfo.ml]: OK for stdout.  See commit 88c2987.

**Comment by xavierleroy** [comment]
> I guess my concern is that I'd like to have a way to be able to use my good old 80x25 terminal.

Just use it, everything should be fine, for several reasons:

* The number of rows N (that is queried with `ioctl` and defaults to 24) can be less than the actual number of rows.  N is used to give up on the fancy underlining if the area to highlight is more than N-2 lines long, meaning that some of it may have scrolled off the top of the screen and reprinting it highlighted would mess the display.  So, if you have 25 rows and OCaml thinks you have 24, the heuristic above will be a tad conservative, but you won't notice.
* There is a very good chance that `ioctl` returns the correct number of columns and rows for your device.  I wouldn't be surprised if the shell derives its `LINES` and `COLUMNS` variables from `ioctl` data.  So, if the former have the correct values, the latter should also have them.
* Once more, the `LINES` variable is for shell script use only: the 4 full-screen applications I tried (Emacs, Vim, top and mutt) just ignore it and determine screen size by other means.

If you tried the original pull request and it didn't work as expected, this is probably due to a bug in my `ioctl` code that confused rows and columns...  Today's version is better.

File: byterun/terminfo.c [hunk 3]
```diff
@@ -13,122 +13,29 @@
...
-{
-  caml_invalid_argument("Terminfo.standout");
-  return Val_unit;
-}
+#ifndef _WIN32
+#include <sys/ioctl.h>
+#endif
 
-CAMLexport value caml_terminfo_resume (value lines)
+CAMLprim value caml_terminfo_rows(value vchan)
```
Comment (dra27) [inline byterun/terminfo.c]: Given that terminfo.c is now reduced to a single Unix-only function, might unix.c be a better place? I say this because there is a Windows implementation of this (for Windows 10+) which I'm happy (and intending, if @nojb doesn't get there first!) to work on, so if win32.c gains a stub version of caml_terminfo_resume for now, then it would be updated later.

File: utils/terminfo.mli [hunk 4]
```diff
@@ -18,9 +18,10 @@
...
   | Bad_term
-  | Good_term of int  (* number of lines of the terminal *)
-;;
-external setup : out_channel -> status = "caml_terminfo_setup";;
-external backup : int -> unit = "caml_terminfo_backup";;
-external standout : bool -> unit = "caml_terminfo_standout";;
-external resume : int -> unit = "caml_terminfo_resume";;
+  | Good_term
+
+val setup : unit -> status
```
Comment (dra27) [inline utils/terminfo.mli]: The old mechanism of caching the channel in C was a bit ugly, but it feels similarly wrong to be hardcoding this to stdout. Given such a "big" change, couldn't this either become a functor or have type `status` converted to a "terminal" or something which gets passed around to the other functions just so that the possibility exists to reuse this for stderr in future?

File: byterun/terminfo.c (hunk 3 above)
Comment (xavierleroy) [reply byterun/terminfo.c]: That's a good suggestion.  We can do this now or later.

File: utils/terminfo.mli (hunk 4 above)
Comment (xavierleroy) [reply utils/terminfo.mli]: Of course the functions from the `Terminfo` module could (should?) take an `out_channel` so that they are not specific to stdout.  However, stdout is hardcoded one level up, in `Location.highlight_terminfo`, which takes a `ppf: Format.formatter` as parameter, but has no way to extract the underlying `out_channel` (if any).  Hence it assumes that `ppf = Format.stdout` and works from there.

**Comment by xavierleroy** [comment]
I just pushed some simplifications and cleanups following @dra27 suggestions.  

Travis seems broken but this branch passed CI "precheck" at Inria.

I am now requesting a formal yes/no review.

**Review by shindere — APPROVED** [review]

**Comment by shindere** [comment]
You may want to update dependencies.
